Add ray tracing module (#16029)
idaholab/moose: modules/ray_tracing/doc/content/source/userobject/RayTracingStudy.md

# RayTracingStudy

The `RayTracingStudy` is the base object for generating and tracing of [Rays](Ray.md) with the [Ray Tracing Module](ray_tracing/index.md). For information about the manipulation of [Rays](Ray.md) once they have began tracing, see [RayKernels/index.md] and [RayBCs/index.md].

!alert tip

> The majority of the information presented here is for +advanced+ used only. The use of this module for simple cases like line sources (see [line_sources.md]) and line integrals (see [line_integrals.md]) utilize the [RepeatableRayStudy.md]. Please see [RepeatableRayStudy.md] first for a more general understanding. If the [RepeatableRayStudy.md] is not sufficient for your use case, please then reference the [RepeatableRayStudyBase.md] instead.

Reviewer: ```suggestion
The majority of the information presented here is for +advanced+ users only. The use of this module for simple cases like line sources (see [line_sources.md]) and line integrals (see [line_integrals.md]) utilize the [RepeatableRayStudy.md]. Please see [RepeatableRayStudy.md] first for a more general understanding. If the [RepeatableRayStudy.md] is not sufficient for your use case, please then reference the [RepeatableRayStudyBase.md] instead.
```

- [Ray Pool](#ray-pool): A shared pool of [Rays](Ray.md) that allows for their reuse after they are called to be destructed
- [Unique Ray IDs](#unique-ray-ids): Generates a unique ID across processors and threads for a [Ray.md]
- [Side Normal Caching](#side-normal-caching): Generates and caches (for future use) outward normals for an element

### Ray Data Registration

The data and auxiliary data associated with a [Ray.md] is stored on the [Ray.md] itself in the form of two `std::vector<RayData>` (`RayData` is typically a `Real`), which are accessed via `Ray::data()` and `Ray::auxData()`. A data registration system exists that allows for the study, [RayKernels/index.md], and [RayBCs/index.md] to request the data and auxiliary data that they need upon construction.

To register value(s), use the methods `registerRayData()` and `registerRayAuxData()`, which return either a single index or a vector of indices into the `Ray::data()` and `Ray::auxData()` vectors that you should utilize. If the same value is registered (uses the same name) within multiple objects, the same index will be returned.

> The `Ray::data()` and `Ray::auxData()` sizes are set automatically under the hood to the registered sizes in some derived studies, but in general for more advanced used cases you should always size the data on each [Ray.md] before it is added to the buffer to be traced. More so if you are modifying the data before a trace - you will likely need to resize the vectors on the [Ray.md] first. The data should be sized using the `rayDataSize()` and `rayAuxDataSize()` methods found in the study. These methods will return the current size of data and auxiliary data registered.

Reviewer: ```suggestion
The `Ray::data()` and `Ray::auxData()` sizes are set automatically under the hood to the registered sizes in some derived studies, but in general for more advanced use cases you should always size the data on each [Ray.md] before it is added to the buffer to be traced. More so if you are modifying the data before a trace---you will likely need to resize the vectors on the [Ray.md] first. The data should be sized using the `rayDataSize()` and `rayAuxDataSize()` methods found in the study. These methods will return the current size of data and auxiliary data registered.
```

The public methods `registeredRayID()` and `registeredRayName()` can be utilized to map a registered [Ray.md] name to its ID and an ID to its registered name, respectively.

### Ray Banking

It is often useful to examine a [Ray.md] after is has completed tracing. An example use case is for obtaining the accumulated integral for a [Ray.md] that is integrating a field along a line (see [line_integrals.md] and [IntegralRayKernel.md]). Within line integrals, the [Ray.md] banks will be accessed to obtain the final accumulated value in a [Ray.md].

With the `_bank_rays_on_completion` private parameter set to true, all [Rays](Ray.md) that complete on a given processor are stored in the private member variable `_ray_bank`. This bank can be accessed through the `rayBank()` method or via `getBankedRay()`, which will return the [Ray.md] with the requested ID on the processor that ended the [Ray.md].

In addition, the methods `getBankedRayData()` and `getBankedRayAuxData()` are available to get a single data value from a [Ray.md] from the banks after it has completed. For these methods, the resulting value is replicated across all processors.

### Ray Pool

> For simulations with the [ray_tracing/index.md] that generate of a large number of [Ray.md] objects, it becomes advantageous to minimize the construction of new [Ray.md] objects. More so whenever the size of the data on the [Ray.md] is significant.

Reviewer: ```suggestion
For simulations with the [ray_tracing/index.md] that generate of a large number of [Ray.md] objects, it becomes advantageous to minimize the construction of new [Ray.md] objects. This is more important whenever the size of the data on the [Ray.md] is significant.
```

Reviewer: What is "significant"?

Author: Man, you and your "small", "large", "significant" comments 😆 (which I approve of)

I have an idea for all of them except for this one.

> [Ray.md] storage is typically wrapped as a `std::shared_ptr<Ray>`. The significance of using a `shared_ptr` is that the destructor is called when the use count reaches zero, that is, when nobody is holding onto the [Ray.md]. The internal tracing algorithm will decrease the use count for a traced [Ray.md] after it has completed tracing. That is, if no other shared ownership of the [Ray.md] exists, the use count will reach zero. This is why [#ray-banking] exists - the [Ray.md] bank will increase the use count of the [Ray.md] so that it is available for use after tracing and is not destructed.

Reviewer: ```suggestion
[Ray.md] storage is typically wrapped as a `std::shared_ptr<Ray>`. The significance of using a `shared_ptr` is that the destructor is called when the use count reaches zero, that is, when nobody is holding onto the [Ray.md]. The internal tracing algorithm will decrease the use count for a traced [Ray.md] after it has completed tracing. That is, if no other shared ownership of the [Ray.md] exists, the use count will reach zero. This is why [#ray-banking] exists---the [Ray.md] bank will increase the use count of the [Ray.md] so that it is available for use after tracing and is not destructed.
```
